cpp/src/merge/merge.cu

```diff
@@ -350,79 +351,125 @@ private:
-    v_merged_cols.reserve(n_cols);
+    std::vector<std::unique_ptr<column>> merged_cols;
+    merged_cols.reserve(n_cols);
+
+    const column_merger merger{merged_indices, mr, stream};
+    transform(left_table.begin(), left_table.end(), 
+             right_table.begin(), 
+             std::back_inserter(merged_cols),
+            [&](auto& left_col, auto& right_col){
+              return cudf::experimental::type_dispatcher(left_col.type(),
+                                                         merger,
+                                                         left_col,
+                                                         right_col);
+            });
+
+    return std::make_unique<cudf::experimental::table>(std::move(merged_cols));
+}
 
-    column_merger merger{merged_indices, mr, stream};
 
-    for(auto i=0;i<n_cols;++i)
-      {
-        const auto& left_col = left_table.column(i);
-        const auto& right_col= right_table.column(i);
-
-        auto merged = cudf::experimental::type_dispatcher(left_col.type(),
-                                                          merger,
-                                                          left_col,
-                                                          right_col);
-        v_merged_cols.emplace_back(std::move(merged));
-      }
+table_ptr_type merge(std::vector<table_view> const& tables_to_merge,
+                     std::vector<cudf::size_type> const& key_cols,
+                     std::vector<cudf::order> const& column_order,
+                     std::vector<cudf::null_order> const& null_precedence,
+                     rmm::mr::device_memory_resource* mr,
+                     cudaStream_t stream = 0) {
+    CUDF_EXPECTS(!tables_to_merge.empty(), "Need to pass at least one table");
+
+    auto const& first_table = tables_to_merge.front();
+    const auto n_cols = first_table.num_columns();
+
+    for (auto const& table: tables_to_merge) {
+      CUDF_EXPECTS(n_cols == table.num_columns(),
+                   "Mismatched number of columns");
+      CUDF_EXPECTS(cudf::have_same_types(first_table, table),
+                   "Mismatched column types");
+    }
+
+    CUDF_EXPECTS(!key_cols.empty(),
+                 "Empty key_cols");
+    CUDF_EXPECTS(key_cols.size() <= static_cast<size_t>(n_cols), 
+                 "Too many values in key_cols");
+
+    CUDF_EXPECTS(key_cols.size() == column_order.size(),
+                 "Mismatched size between key_cols and column_order");
+    CUDF_EXPECTS(column_order.size() <= static_cast<size_t>(n_cols), 
+                 "Too many values in column_order");
+
+    if (tables_to_merge.size() == 1 || n_cols == 0) {
+      return std::make_unique<cudf::experimental::table>(first_table);
+    }
 
-    return std::make_unique<cudf::experimental::table>(std::move(v_merged_cols));
+    // A queue of (table view, table) pairs
+    std::deque<std::pair<table_view, table_ptr_type>> merge_queue;
```

Comment:
What's the purpose of storing a pair of a table and view to the table? `table` is implicitly convertible to a `table_view`, so I would think there isn't a need to store it's corresponding `table_view`.

Reply:
The reason for storing both views and tables is to be able to uniformly process both the input tables (not owned by this function) and the intermediate results (owned). By storing a unique_ptr of the table in the same struct as the view, I ensure that the intermediate tables are deleted as soon as struct is removed from the queue. For the input table views we don't need to deallocate the tables, so the corresponding ptr is null.
I can declare a struct instead of using a pair here, good struct & member naming could help make this code clearer.

Reply:
Ah... okay, that makes sense.

Reply:
Related question, I was wondering if it made sense to use an ordered container like a `set` or a `priority_queue` and order the tables to merge based on the number of rows such that we're always merging the two smallest tables together first.

Reply:
We find ourselves some times storing both views and tables. I wonder if we dont need an abstraction for something that is a view OR a table and can release resources in the case of ownership.

Reply:
> Related question, I was wondering if it made sense to use an ordered container like a `set` or a `priority_queue` and order the tables to merge based on the number of rows such that we're always merging the two smallest tables together first.

I actually thought of the same thing over the weekend :) We always have N-1 merges, so delaying the merges of the largest tables as much as possible should be the optimal solution. I'll make this change and add a benchmark to verify the performance improvement.

Reply:
In practice for our initial use case, won't the partitions be mostly the same size?

Reply:
> In practice for our initial use case, won't the partitions be mostly the same size?

Ideally, but it can depend significantly on the data skew.

Reply:
> Ideally, but it can depend significantly on the data skew.
The merge benchmark I'm adding has a normal distribution, I think this should mimic the intended use case well.